File: src/ai/providers/deepseek.ts
/**
 * DeepSeek provider adapter.
 * Supports DeepSeek V3.2 (chat) and DeepSeek Reasoner (thinking).
 *
 * Endpoint: https://api.deepseek.com/chat/completions
 * Auth:     Authorization: Bearer <key>
 * Format:   OpenAI-compatible, but reasoning models emit a
 *           reasoning_content field in the delta for thinking tokens.
 * History:  Strip reasoning_content from history messages to avoid
 *           confusing the model on subsequent turns.
 */

import type { ChatMessage, ChatOptions, StreamEvent } from "../types";
import { BaseProvider } from "./base";

const ENDPOINT = "https://api.deepseek.com/chat/completions";

export class DeepSeekProvider extends BaseProvider {
  async chat(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): Promise<ChatMessage> {
    const body = this.buildRequestBody(model, messages, options, false);

    const response = await fetch(ENDPOINT, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(
        `DeepSeek request failed (${response.status}): ${errorText}`
      );
    }

    const data = await response.json();
    const choice = data.choices?.[0];
    const message = choice?.message;

    return {
      role: "assistant",
      content: message?.content ?? "",
      thinking: message?.reasoning_content || undefined,
      timestamp: Date.now(),
    };
  }

  async *chatStream(
    model: string,
    messages: ChatMessage[],
    options?: ChatOptions
  ): AsyncGenerator<StreamEvent> {
    const body = this.buildRequestBody(model, messages, options, true);

    const response = await fetch(ENDPOINT, {
      method: "POST",
      headers: this.buildHeaders(),
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      const errorText = await response.text();
      yield {
        type: "error",
        message: `DeepSeek stream failed (${response.status}): ${errorText}`,
      };
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) {
      yield { type: "error", message: "No response body from DeepSeek" };
      return;
    }

    let inputTokens = 0;
    let outputTokens = 0;

    for await (const chunk of this.parseSSE(reader)) {
      const data = chunk as Record<string, unknown>;

      // Track usage if present
      if (data.usage) {
        const usage = data.usage as Record<string, number>;
        inputTokens = usage.prompt_tokens ?? 0;
        outputTokens = usage.completion_tokens ?? 0;
      }

      const choices = data.choices as Array<Record<string, unknown>> | undefined;
      if (!choices || choices.length === 0) continue;

      const delta = choices[0].delta as Record<string, string> | undefined;
      if (!delta) continue;

      // DeepSeek Reasoner emits reasoning_content for thinking tokens
      if (delta.reasoning_content) {
        yield { type: "thinking", content: delta.reasoning_content };
      }

      if (delta.content) {
        yield { type: "content", content: delta.content };
      }
    }

    yield {
      type: "done",
      usage: { input_tokens: inputTokens, output_tokens: outputTokens },
    };
  }

  private buildHeaders(): Record<string, string> {
    return {
      "Content-Type": "application/json",
      Authorization: `Bearer ${this.apiKey}`,
    };
  }

  private buildRequestBody(
    model: string,
    messages: ChatMessage[],
    options: ChatOptions | undefined,
    stream: boolean
  ): Record<string, unknown> {
    const body: Record<string, unknown> = {
      model,
      messages: this.formatMessages(messages),
      stream,
    };

    if (stream) {
      body.stream_options = { include_usage: true };
    }

    if (options?.maxTokens) {
      body.max_tokens = options.maxTokens;
    }

    if (options?.temperature !== undefined) {
      body.temperature = options.temperature;
    }

    return body;
  }

  /**
   * Format messages for the DeepSeek API.
   * Strips reasoning_content / thinking from assistant messages in history
   * to prevent the model from being confused by prior reasoning traces.
   */
  private formatMessages(
    messages: ChatMessage[]
  ): Array<Record<string, unknown>> {
    return messages.map((m) => {
      const formatted: Record<string, unknown> = {
        role: m.role,
        content: m.content,
      };

      // Do NOT forward thinking/reasoning_content in history — DeepSeek docs
      // specify that reasoning_content must be stripped from assistant turns
      // in multi-turn conversations.

      return formatted;
    });
  }
}

File: crates/provider-adapters/src/deepseek.rs
use serde_json::{json, Value};

use super::types::{ChatRequest, ChatResponse, StreamEvent, Usage};

const DEEPSEEK_API_URL: &str = "https://api.deepseek.com/chat/completions";

/// Build the DeepSeek request body. DeepSeek uses an OpenAI-compatible format.
fn build_request_body(request: &ChatRequest) -> Value {
    let messages: Vec<Value> = request
        .messages
        .iter()
        .map(|msg| {
            json!({
                "role": msg.role,
                "content": msg.content,
            })
        })
        .collect();

    json!({
        "model": request.model,
        "messages": messages,
        "max_tokens": request.max_tokens,
        "temperature": request.temperature,
        "stream": request.stream,
    })
}

/// Send a non-streaming request to the DeepSeek API.
pub async fn send_request(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<ChatResponse, String> {
    let mut body = build_request_body(request);
    body["stream"] = json!(false);

    let response = client
        .post(DEEPSEEK_API_URL)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("DeepSeek request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!("DeepSeek API error ({}): {}", status, text));
    }

    let data: Value = response
        .json()
        .await
        .map_err(|e| format!("Failed to parse DeepSeek response: {}", e))?;

    let content = data["choices"][0]["message"]["content"]
        .as_str()
        .unwrap_or("")
        .to_string();

    // DeepSeek may include reasoning_content for its thinking models
    let thinking = data["choices"][0]["message"]["reasoning_content"]
        .as_str()
        .map(|s| s.to_string());

    let input_tokens = data["usage"]["prompt_tokens"].as_u64().unwrap_or(0) as u32;
    let output_tokens = data["usage"]["completion_tokens"].as_u64().unwrap_or(0) as u32;

    Ok(ChatResponse {
        content,
        thinking,
        usage: Usage {
            input_tokens,
            output_tokens,
        },
    })
}

/// Send a streaming request to the DeepSeek API.
pub async fn send_stream(
    client: &reqwest::Client,
    api_key: &str,
    request: &ChatRequest,
) -> Result<reqwest::Response, String> {
    let mut body = build_request_body(request);
    body["stream"] = json!(true);

    let response = client
        .post(DEEPSEEK_API_URL)
        .header("Authorization", format!("Bearer {}", api_key))
        .header("Content-Type", "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|e| format!("DeepSeek stream request failed: {}", e))?;

    if !response.status().is_success() {
        let status = response.status();
        let text = response
            .text()
            .await
            .unwrap_or_else(|_| "unknown error".to_string());
        return Err(format!("DeepSeek stream API error ({}): {}", status, text));
    }

    Ok(response)
}

/// Parse an SSE chunk from DeepSeek's streaming format into unified StreamEvents.
/// DeepSeek uses the same OpenAI-compatible SSE format.
pub fn parse_stream_chunk(chunk: &str) -> Vec<StreamEvent> {
    let mut events = Vec::new();

    for line in chunk.lines() {
        let line = line.trim();
        if !line.starts_with("data: ") {
            continue;
        }
        let data = &line[6..];
        if data == "[DONE]" {
            events.push(StreamEvent {
                event_type: "done".to_string(),
                content: None,
                usage: None,
                message: None,
            });
            continue;
        }

        if let Ok(parsed) = serde_json::from_str::<Value>(data) {
            // Content delta
            if let Some(content) = parsed["choices"][0]["delta"]["content"].as_str() {
                if !content.is_empty() {
                    events.push(StreamEvent {
                        event_type: "content".to_string(),
                        content: Some(content.to_string()),
                        usage: None,
                        message: None,
                    });
                }
            }

            // Reasoning content (DeepSeek thinking)
            if let Some(reasoning) = parsed["choices"][0]["delta"]["reasoning_content"].as_str() {
                if !reasoning.is_empty() {
                    events.push(StreamEvent {
                        event_type: "thinking".to_string(),
                        content: Some(reasoning.to_string()),
                        usage: None,
                        message: None,
                    });
                }
            }

            // Usage
            if let Some(usage_obj) = parsed.get("usage") {
                if !usage_obj.is_null() {
                    let input_tokens = usage_obj["prompt_tokens"].as_u64().unwrap_or(0) as u32;
                    let output_tokens =
                        usage_obj["completion_tokens"].as_u64().unwrap_or(0) as u32;
                    events.push(StreamEvent {
                        event_type: "usage".to_string(),
                        content: None,
                        usage: Some(Usage {
                            input_tokens,
                            output_tokens,
                        }),
                        message: None,
                    });
                }
            }
        }
    }

    events
}
